### ISSUE TITLE
Major SQL injection vulnerability in /api/tsdb/query

### ISSUE THREAD

**Issue by KensingtonTech** [grafana/grafana#32043]
Using Postgresql data source, the `/api/tsdb/query` call used by dashboards embeds raw SQL into the POST data.  This is a major SQL injection vulnerability.

Though this has been previously reported both here and in the community, the issue raised was closed due to it being in the form of a "question", but IMHO this should be urgently revisited.  This is also why I am reporting here as opposed to the vulnerability email address -- it is already known.

## The documentation guidance on SELECT-only users is insufficient to mitigate!

The documented guidance to restrict the data source user to SELECT-only permission doesn't nearly address the scope of the vulnerability.

Even with SELECT-only data source permissions, users would still have virtually unrestricted access to data in a table, which would permit them to dump the contents of an entire table through the Grafana API.  To say nothing of the query / performance havoc that could potentially be wrought by malicious users.

**What you expected to happen**:  Not to pass SQL data from the browser to the data source in dashboards to be used in queries, except when in the query editor.  The Grafana server should handle all panel / variable queries on behalf of the user.

**How to reproduce it (as minimally and precisely as possible)**:  

1.  Create any dashboard with Postgresql in a panel query,
2.  In the browser debugger, watch the POST to `/api/tsdb/query, and look at the request payload.

![image](https://user-images.githubusercontent.com/15734900/111381736-77dd7180-867c-11eb-8fa6-8ed9ef3c76c4.png)

**Anything else we need to know?**:  

**Environment**:
- Grafana version: 7.5.0 beta 1
- Data source type & version: Postgresql
- OS Grafana is installed on: Grafana docker image
- User OS & Browser: MacOS, Chrome 89
- Grafana plugins: N/A
- Others: N/A

Thank you in advance for reconsidering this.

**Comment by torkelo** [comment]
Do you think we should update the docs to make this more clear? There is currently no way to fix this, queries are built dynamically in the browser so we have to send the full query through the API. This is how it works for every Grafana data source.
 
There is one way to fix it but it would require disabling template variables and just use the query saved in the dashboard but template variables are a very heavily used feature that most dashboards take advantage of. 

We do recognize that this is big problem, especially for SQL data sources. We want to find some solution long-term but won't happen any time soon, maybe for v9 next year. I fear it's going to take a lot of work. And with template variables, it will not be 100% secure as validating that variable values coming from the browser have values that the user is allowed to set is close to impossible.

**Comment by septatrix** [comment]
I agree and think that this is a major security drawback. For some data sources this could be partially mitigated e.g. for SQL one might create an SQL view with a limit and one the relevant fields however this is hard to maintain and not applicable to all sources or use cases.

> There is one way to fix it but it would require disabling template variables and just use the query saved in the dashboard but template variables are a very heavily used feature that most dashboards take advantage of.
> 
> We do recognize that this is big problem, especially for SQL data sources. We want to find some solution long-term but won't happen any time soon, maybe for v9 next year. I fear it's going to take a lot of work. And with template variables, it will not be 100% secure as validating that variable values coming from the browser have values that the user is allowed to set is close to impossible.

Why not just send the query arguments separate from the query to the backend? It could then verify whether the query exists on the dashboard and afterwards fill in the arguments. Even if the arguments themselves are not validated this would be a HUGE improvement. Most data sources then support parameterized queries which could be further used to prevent code injection. As this is also used by Grafana in many places the logic should already exists. If there is any data source which does not support parameterized queries (of which I currently cannot think of) it can still fall back to building the string using `fmt.Sprintf` which is at least as safe as the current implementation.

**Comment by a-imran99** [comment]
> We do recognize that this is big problem, especially for SQL data sources. We want to find some solution long-term but won't happen any time soon, maybe for v9 next year.

is this feature of securing the rawSQL brought in or fixed in Grafana version 9 which got released this year. @torkelo Or is it still being looked into? It would be nice if replied. Thank you.

**Comment by septatrix** [comment]
> is this feature of securing the rawSQL brought in or fixed in Grafana version 9 which got released this year. @torkelo 

To my knowledge sadly no

> Or is it still being looked into? It would be nice if replied. Thank you.

Given the lack of any response I fear the answer is no here too though an official response would definitely be much appreciated.

**Comment by a-imran99** [comment]
Hey thanks @septatrix , For now we have just reversed the SQL query string from UI and then again re-reversed the same at the API endpoint logic. We know that it is not a complete solution but we hope that atleast it'll trick the testing tool that no SQL query is sent in request payload.

**Comment by codeincarnate** [comment]
This is something we're definitely looking into. The context around this has changed a fair amount and the fundamentals of this should be handled with [Grafana Public Dashboards](https://grafana.com/docs/grafana/latest/dashboards/dashboard-public/) related code. That being said it may take some time to get to a final solution.

**Comment by TimUnderhay** [comment]
It would be great to get info on the present state of this.

**Comment by jahtux** [comment]
It would be really useful to get info on the present state of this. This is a stopped security risk to implement Grafana in any organization. Thanks!

**Comment by septatrix** [comment]
Still relevant for security

**Comment by sinhpn92** [comment]
Hello everyone,
Have any update to fix this issue? It is huge security issue :(
